account_asset_management/models/account_asset_profile.py

```diff
@@ -0,0 +1,180 @@
+             "of this profile when created by invoices.")
+    asset_product_item = fields.Boolean(
+        string='Create an asset by product item',
+        help="By default during the validation of an invoice, an asset "
+             "is created by invoice line as long as an accounting entry is "
+             "created by invoice line. "
+             "With this setting, an accounting entry will be created by "
+             "product item. So, there will be an asset by product item.")
+    active = fields.Boolean(default=True)
+
+    @api.model
+    def _default_company_id(self):
+        return self.env['res.company']._company_default_get('account.asset')
+
+    @api.model
+    def _selection_method(self):
+        return[
+            ('linear', _('Linear')),
+            ('linear-limit', _('Linear up to Salvage Value')),
+            ('degressive', _('Degressive')),
+            ('degr-linear', _('Degressive-Linear')),
+            ('degr-limit', _('Degressive  up to Salvage Value')),
+        ]
+
+    @api.model
+    def _selection_method_period(self):
+        return [
+            ('month', _('Month')),
+            ('quarter', _('Quarter')),
+            ('year', _('Year')),
+        ]
+
+    @api.model
+    def _selection_method_time(self):
+        """
+        Install the 'account_asset_management_method_number_end' to enable the
+        'Number' and 'End' Time Methods.
+        """
+        return [
+            ('year', _('Number of Years')),
+        ]
+
+    @api.multi
+    @api.constrains('method')
+    def _check_method(self):
+        for profile in self:
+            if profile.method == 'degr-linear' and \
+                    profile.method_time != 'year':
+                raise UserError(
+                    _("Degressive-Linear is only supported for Time Method = "
+                      "Year."))
+
+    @api.onchange('method_time')
+    def _onchange_method_time(self):
+        if self.method_time != 'year':
+            self.prorata = True
+
+    @api.model
+    def create(self, vals):
+        if vals.get('method_time') != 'year' and not vals.get('prorata'):
+            vals['prorata'] = True
+        profile = super(AccountAssetProfile, self).create(vals)
+        acc_id = vals.get('account_asset_id')
+        if acc_id:
+            account = self.env['account.account'].browse(acc_id)
+            if not account.asset_profile_id:
+                account.write({'asset_profile_id': profile.id})
+        return profile
+
+    @api.multi
+    def write(self, vals):
+        if vals.get('method_time'):
+            if vals['method_time'] != 'year' and not vals.get('prorata'):
+                vals['prorata'] = True
+        super(AccountAssetProfile, self).write(vals)
+        for profile in self:
+            acc_id = vals.get('account_asset_id')
+            if acc_id:
+                account = self.env['account.account'].browse(acc_id)
+                if not account.asset_profile_id:
```

Comment:
Lines 175-178 could be replaced by `if not self.env['account.account'].browse(vals.get('account_asset_id'))`.
This shouldn't crash.

Reply:
Actually this is a really buggy behaviour, I did change and added a TODO.

Reply:
I think the 'random' behaviour in the TODO is really highly theoretical, since the 'account_asset_id' field is only set via the GUI hence we should have only one recordset.
The old 'for profile in self' code was more safe than the self[-1] now.
We could add a raise on 'if account' when len(self) != 1 and remove the TODO and replace self[-1] by self.

Reply:
@luc-demeyer If there's no extension I agree this can only be changed on GUI.
But the model can be extended and restricting the update of multiple records (i.e. raising an error if `len(self) > 1` on `write`) at the same time is not a good practice.
Until now, it did happen that the last `account.profile` in self is defined as `asset_profile_id`, so we should keep this behaviour if we don't have a more specific requirement of what should happen.
The new code only does it more effectively than looping on the records.

IMO that could also be discussed after this PR gets merged, let's keep it a straight migration here.